File: sdk/pythonexamples/transaction_aggregate.py
#!/usr/bin/env python

#
# Reads utf8 encoded files with names "part<id>.txt".
# Creates aggregate transaction where each files is a message inside embedded transfers.
# Signs aggregate transaction using private key provided in file specified via `--private` switch.
#

import argparse
from binascii import unhexlify
from pathlib import Path

import sha3

from symbolchain.core.CryptoTypes import Hash256, PrivateKey
from symbolchain.core.facade.SymFacade import SymFacade
from symbolchain.core.sym.KeyPair import KeyPair
from symbolchain.core.sym.MerkleHashBuilder import MerkleHashBuilder


def read_private_key(private_filename):
    return KeyPair(PrivateKey(unhexlify(read_contents(private_filename).strip())))


def read_contents(filepath):
    with open(filepath, 'rt', encoding='utf8') as infile:
        return infile.read()


def add_embedded_transfers(facade, public_key):
    # obtain recipient from public_key, so direct all transfers to 'self'
    embedded_transactions = []
    recipient = facade.network.public_key_to_address(public_key)
    resources = Path(__file__).parent / 'resources'
    for filepath in sorted(resources.glob('part*.txt')):
        msg = read_contents(filepath)
        embedded = facade.transaction_factory.create_embedded({
            'type': 'transfer',
            'signer_public_key': public_key,
            'recipient_address': recipient,
            # note: additional 0 byte at the beginning is added for compatibility with explorer
            # and other tools that treat messages starting with 00 byte as "plain text"
            'message': bytes(1) + msg.encode('utf8')
        })

        embedded_transactions.append(embedded)

        print('----> {} length in bytes: {}'.format(filepath.name, len(embedded.message)))

    return embedded_transactions


def main():
    parser = argparse.ArgumentParser(description='create aggregate')
    parser.add_argument('--private', help='path to file with private key', required=True)
    args = parser.parse_args()

    facade = SymFacade('public_test')
    key_pair = read_private_key(args.private)

    embedded_transactions = add_embedded_transfers(facade, key_pair.public_key)
    hash_builder = MerkleHashBuilder()
    for embedded_transaction in embedded_transactions:
        embedded_transaction_hash = sha3.sha3_256(embedded_transaction.serialize()).digest()
        hash_builder.update(Hash256(embedded_transaction_hash))

    merkle_hash = hash_builder.final()

    aggregate = facade.transaction_factory.create({
        'type': 'aggregateComplete',
        'signer_public_key': key_pair.public_key,
        'fee': 0,
        'deadline': 1,
        'transactions_hash': merkle_hash,
        'transactions': embedded_transactions
    })

    signature = facade.sign_transaction(key_pair, aggregate)
    aggregate.signature = signature.bytes

    print(aggregate)


if __name__ == '__main__':
    main()

File: sdk/java/examples/src/main/java/org/symbol/examples/TransactionAggregate.java
package org.symbol.examples;

import java.nio.charset.StandardCharsets;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import org.symbol.sdk.CryptoTypes;
import org.symbol.sdk.facade.SymbolFacade;
import org.symbol.sdk.symbol.KeyPair;
import org.symbol.sdk.symbol.SymbolTransactionFactory;
import org.symbol.sdk.symbol.descriptors.AggregateCompleteTransactionV3Descriptor;
import org.symbol.sdk.symbol.descriptors.TransferTransactionV1Descriptor;
import org.symbol.sdk.symbol.descriptors.UnresolvedMosaicDescriptor;
import org.symbol.sdk.symbol.models.Cosignature;
import org.symbol.sdk.symbol.models.EmbeddedTransaction;
import org.symbol.sdk.symbol.models.EmbeddedTransferTransactionV1;
import org.symbol.sdk.symbol.models.Transaction;
import org.symbol.sdk.utils.ArrayHelpers;

/**
 * Reads every bundled {@code part*} resource file and packages them as embedded transfer transactions inside a single Symbol
 * aggregate-complete transaction signed by the key from the {@code --private} file.
 */
public final class TransactionAggregate {
	private TransactionAggregate() {
	}

	private static List<EmbeddedTransaction> addEmbeddedTransfers(final SymbolFacade facade, final CryptoTypes.PublicKey publicKey,
			final List<Path> partFiles) {
		// obtain recipient from publicKey, so direct all transfers to 'self'
		final var recipientAddress = facade.network.publicKeyToAddress(publicKey);

		final List<EmbeddedTransaction> result = new ArrayList<>();
		for (final Path filename : partFiles) {
			final String message = ExamplesUtils.readContents(filename);
			final byte[] messageBytes = message.getBytes(StandardCharsets.UTF_8);
			// note: additional 0 byte at the beginning is added for compatibility with explorer
			// and other tools that treat messages starting with 00 byte as "plain text"
			final byte[] prefixed = ArrayHelpers.concat(new byte[1], messageBytes);

			final TransferTransactionV1Descriptor descriptor = new TransferTransactionV1Descriptor(recipientAddress, (List<UnresolvedMosaicDescriptor>) null, prefixed);
			final EmbeddedTransferTransactionV1 embeddedTransaction = (EmbeddedTransferTransactionV1) facade
					.createEmbeddedTransactionFromTypedDescriptor(descriptor, publicKey);

			System.out.println("----> " + filename.getFileName() + " length in bytes: " + embeddedTransaction.getMessage().length);
			result.add(embeddedTransaction);
		}

		return result;
	}

	public static void main(final String[] args) {
		final String privateArg = ExamplesUtils.parseFlag(args, "--private");
		if (null == privateArg)
			throw new IllegalArgumentException("missing required --private <path-to-private-key>");

		final Path privatePath = Paths.get(privateArg);

		final SymbolFacade facade = new SymbolFacade("testnet");
		final KeyPair keyPair = ExamplesUtils.readPrivateKey(privatePath);

		final List<Path> partFiles = ExamplesUtils.findBundledFiles("part", ".txt");
		final List<EmbeddedTransaction> embeddedTransactions = addEmbeddedTransfers(facade, keyPair.getPublicKey(), partFiles);
		final CryptoTypes.Hash256 merkleHash = SymbolFacade.hashEmbeddedTransactions(embeddedTransactions);

		final AggregateCompleteTransactionV3Descriptor aggregateDescriptor = new AggregateCompleteTransactionV3Descriptor(merkleHash,
				embeddedTransactions, (List<Cosignature>) null);
		// pin deterministic header values in the descriptor (like the JS / Python examples) so the output stays comparable
		final Map<String, Object> aggregateMap = aggregateDescriptor.toMap();
		aggregateMap.put("signerPublicKey", keyPair.getPublicKey());
		aggregateMap.put("fee", 0L);
		aggregateMap.put("deadline", 1L);
		final Transaction aggregateTransaction = facade.transactionFactory.create(aggregateMap);

		final CryptoTypes.Signature signature = facade.signTransaction(keyPair, aggregateTransaction);
		SymbolTransactionFactory.attachSignature(aggregateTransaction, signature);

		System.out.println("Hash: " + facade.hashTransaction(aggregateTransaction) + "\n");
		System.out.println(aggregateTransaction.toString());
	}
}
